File: app/services/auth_service.py
from sqlalchemy.orm import Session
from sqlalchemy.exc import IntegrityError
from app.schemas.user import UserCreate, UserLogin, Token
from app.models.user import User
from app.core.security import verify_password, create_access_token, get_password_hash
from app.core.config import settings
from datetime import datetime, timedelta

class AuthService:
    def __init__(self, db: Session):
        self.db = db

    async def register_user(self, user_data: UserCreate) -> dict:
        """Đăng ký user mới"""
        # Kiểm tra email đã tồn tại
        existing_user = self.db.query(User).filter(User.email == user_data.email).first()
        if existing_user:
            raise ValueError("Email already registered")

        # Kiểm tra username đã tồn tại
        existing_username = self.db.query(User).filter(User.username == user_data.username).first()
        if existing_username:
            raise ValueError("Username already taken")

        # Tạo user mới
        hashed_password = get_password_hash(user_data.password)
        db_user = User(
            username=user_data.username,
            email=user_data.email,
            password_hash=hashed_password,
            full_name=user_data.full_name,
            date_of_birth=user_data.date_of_birth,
            learning_level=user_data.learning_level or "Beginner",
            preferred_language=user_data.preferred_language or "vi"
        )

        try:
            self.db.add(db_user)
            self.db.commit()
            self.db.refresh(db_user)
            
            return {
                "message": "User registered successfully",
                "user_id": str(db_user.user_id),
                "username": db_user.username
            }
        except IntegrityError:
            self.db.rollback()
            raise ValueError("Registration failed due to data conflict")

    async def authenticate_user(self, username: str, password: str) -> Token:
        """Xác thực user và tạo token"""
        user = self.db.query(User).filter(User.username == username).first()
        
        if not user or not verify_password(password, user.password_hash):
            raise ValueError("Invalid username or password")
        
        if not user.is_active:
            raise ValueError("User account is disabled")

        access_token_expires = timedelta(minutes=settings.ACCESS_TOKEN_EXPIRE_MINUTES)
        access_token = create_access_token(
            data={"sub": str(user.user_id)}, 
            expires_delta=access_token_expires
        )

        return Token(
            access_token=access_token,
            token_type="bearer",
            expires_in=settings.ACCESS_TOKEN_EXPIRE_MINUTES * 60
        )

    async def refresh_token(self, token: str) -> Token:
        """Làm mới access token"""
        # Implement token refresh logic here
        # For now, just return a new token
        access_token_expires = timedelta(minutes=settings.ACCESS_TOKEN_EXPIRE_MINUTES)
        access_token = create_access_token(
            data={"sub": "user_id"}, 
            expires_delta=access_token_expires
        )

        return Token(
            access_token=access_token,
            token_type="bearer",
            expires_in=settings.ACCESS_TOKEN_EXPIRE_MINUTES * 60
        )

File: fontend/src/services/authService.js
/**
 * Service Xác Thực
 * 
 * Service này xử lý tất cả các API liên quan đến xác thực:
 * - Đăng nhập (login)
 * - Đăng ký (register) 
 * - Đăng xuất (logout)
 * - Refresh token
 * - Verify token
 * 
 * Trong development sử dụng mock data, production sử dụng real API
 */

import api from "./api"
import { API_ENDPOINTS } from "../config/api.config"

// Sử dụng mock service trong development, real API trong production

export const authService = {
  // API đăng nhập - xác thực email và password
  login: async (identifier, password) => {
    try {
      const response = await api.post(API_ENDPOINTS.AUTH.LOGIN, { identifier, password })
      return {
        user: response.user,
        token: response.access_token,
        expiresIn: response.expires_in,
      }
    } catch (error) {
      const message = error?.detail || error?.message || "Login failed"
      throw new Error(message)
    }
  },
  // API đăng ký - tạo tài khoản mới
  register: async (userData) => {
    try {
      const response = await api.post(API_ENDPOINTS.AUTH.REGISTER, userData)
      return {
        user: response,
        token: null,
      }
    } catch (error) {
      const message = error?.detail || error?.message || "Register failed"
      throw new Error(message)
    }
  },

  // API đăng xuất - xóa session
  logout: () =>api.post(API_ENDPOINTS.AUTH.LOGOUT),

  // API refresh token - làm mới access token
  refreshToken: () =>api.post(API_ENDPOINTS.AUTH.REFRESH),

  // API verify token - kiểm tra token có hợp lệ không
  verifyToken: () =>api.get(API_ENDPOINTS.AUTH.VERIFY),
}
